tests/codegen/intrinsics/transmute.rs

```diff
@@ -76,23 +80,80 @@ pub unsafe fn check_from_uninhabited(x: BigNever) -> u16 {
     }
 }
 
+// CHECK-LABEL: @check_intermediate_passthrough(
+#[no_mangle]
+pub unsafe fn check_intermediate_passthrough(x: u32) -> i32 {
+    // CHECK: start
+    // CHECK: %[[TMP:.+]] = add i32 1, %x
+    // CHECK: %[[RET:.+]] = add i32 %[[TMP]], 1
+    // CHECK: ret i32 %[[RET]]
+    unsafe {
+        transmute::<u32, i32>(1 + x) + 1
+    }
+}
+
+// CHECK-LABEL: @check_nop_pair(
+#[no_mangle]
+pub unsafe fn check_nop_pair(x: (u8, i8)) -> (i8, u8) {
+    // CHECK-NOT: alloca
+    // CHECK: %0 = insertvalue { i8, i8 } poison, i8 %x.0, 0
+    // CHECK: %1 = insertvalue { i8, i8 } %0, i8 %x.1, 1
+    // CHECK: ret { i8, i8 } %1
+    unsafe {
+        transmute(x)
+    }
+}
+
 // CHECK-LABEL: @check_to_newtype(
 #[no_mangle]
 pub unsafe fn check_to_newtype(x: u64) -> Scalar64 {
-    // CHECK: %0 = alloca i64
-    // CHECK: store i64 %x, ptr %0
-    // CHECK: %1 = load i64, ptr %0
-    // CHECK: ret i64 %1
+    // CHECK-NOT: alloca
+    // CHECK: ret i64 %x
     transmute(x)
 }
 
 // CHECK-LABEL: @check_from_newtype(
 #[no_mangle]
 pub unsafe fn check_from_newtype(x: Scalar64) -> u64 {
-    // CHECK: %0 = alloca i64
-    // CHECK: store i64 %x, ptr %0
-    // CHECK: %1 = load i64, ptr %0
-    // CHECK: ret i64 %1
+    // CHECK-NOT: alloca
+    // CHECK: ret i64 %x
+    transmute(x)
+}
+
+// CHECK-LABEL: @check_aggregate_to_bool(
+#[no_mangle]
+pub unsafe fn check_aggregate_to_bool(x: Aggregate8) -> bool {
+    // CHECK: %x = alloca %Aggregate8, align 1
+    // CHECK: %[[BYTE:.+]] = load i8, ptr %x, align 1
+    // CHECK: %[[BOOL:.+]] = trunc i8 %[[BYTE]] to i1
```

Comment:
Why does this produce a `trunc`? Intuitively I'd think that calling `check_aggregate_to_bool` with anything but `0` or `1` is UB, so we don't have to truncate here, after all transmute should just be an noop 🤔

Reply:
Yeah, it should be a nop.  The problem is that they're different LLVM types, so it needs *some* instruction in the IR.

For example, even <https://rust.godbolt.org/z/KxYvhsEen>
```
pub fn foo(b: &bool) -> bool {
    *b
}
```
compiles to something that needs an instruction (the `trunc` turns into an `icmp`):
```llvm
define noundef zeroext i1 @_ZN7example3foo17h89a4116ff69f8f47E(ptr noalias nocapture noundef readonly align 1 dereferenceable(1) %b) unnamed_addr #0 !dbg !6 {
  %0 = load i8, ptr %b, align 1, !dbg !11, !range !12, !noundef !10
  %1 = icmp ne i8 %0, 0, !dbg !11
  ret i1 %1, !dbg !13
}
```

Hopefully we'll get [`trunc nuw i8 %b to i1`](https://discourse.llvm.org/t/best-ir-for-mixed-width-shift/66144/2?u=scottmcm) one day to more directly represent the "look, it's a nop" thing.

(The codegen tests with `bool` are here because my first attempt at this didn't have the [to/from immediate](https://github.com/rust-lang/rust/pull/109843/files#diff-db08b7c8b00530c7183cf2e42f25dc93b02da93fb40edadbd009eee6ad60e3a1R231) parts and thus blew up horribly because the LLVM IR didn't match up on its types.)

Reply:
Okay, so the answer is "llvm doesn't support good options here (yet?)", I see.